### PR TITLE
Implement Bullseye Polytope Clean Label Attack

### LINKED ISSUE #831: Implement Convex Polytope clean label attack
**Is your feature request related to a problem? Please describe.**
This attack allows for transferability of clean label attacks.

**Describe the solution you'd like**
Implementation of the paper Transferable Clean-Label Poisoning Attacks on Deep Neural Nets 
https://arxiv.org/abs/1905.05897

**Describe alternatives you've considered**
NA

**Additional context**
NA

@ebubae

### PR DESCRIPTION
**Pull request by ebubae**
# Description

Implementation of [Bullseye Polytope: A Scalable Clean-Label Poisoning Attack with Improved Transferability](https://arxiv.org/abs/2005.00191) (Aghakhani, et. al. 2020). Based on the implementation in https://github.com/ucsb-seclab/BullseyePoison. 

Fixes #831 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

- [x] test_bullseye_polytope_attack.py

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class BullseyePolytopeAttack(PoisoningAttackWhiteBox):
+    """
+    Implementation of Bullseye Polytope Attack by Aghakhani, et. al. 2020.
+    "Bullseye Polytope: A Scalable Clean-Label Poisoning Attack with Improved Transferability"
+
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:48]:
```suggestion
```

File: tests/attacks/poison/test_bullseye_polytope_attack.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline tests/attacks/poison/test_bullseye_polytope_attack.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: tests/attacks/poison/test_bullseye_polytope_attack.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+@pytest.mark.skipMlFramework("non_dl_frameworks", "tensorflow", "mxnet", "keras", "kerastf")
+def test_poison(art_warning, get_default_mnist_subset, image_dl_estimator):
+    try:
+        (x_train, y_train), (_, _) = get_default_mnist_subset
+        classifier, _ = image_dl_estimator(functional=True)
+        target = np.expand_dims(x_train[3], 0)
+        attack = BullseyePolytopeAttack(classifier, target, len(classifier.layer_names) - 2)
+        poison_data, poison_labels = attack.poison(x_train[5:10], y_train[5:10])
+
```
Comment (beat-buesser) [inline tests/attacks/poison/test_bullseye_polytope_attack.py:39]: Should we add a test for checking that samples in `poison_data` are different from `x_train`?

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        # We can try enforcing the convex polytope in one of the multiple realizations of the feature,
+        # but empirically one realization is enough.
+        target_feat_list = []
+        # Coefficients for the convex combination.
+        # Initializing from the coefficients of last step gives faster convergence.
+        s_init_coeff_list = []
+        n_poisons = len(x)
+        for n, net in enumerate(self.subsistute_networks):
+            # net = net.model
+            # net.eval()
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:181]: Can we remove these lines?

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+            optimizer = torch.optim.SGD(poison_batch.parameters(), lr=self.learning_rate, momentum=self.momentum)
+        elif opt_method == 'signedadam':
+            logger.info("Using Signed Adam to craft poison samples")
+            optimizer = SignedAdam(poison_batch.parameters(), lr=self.learning_rate, betas=(self.momentum, 0.999))
+        elif opt_method == 'adam':
+            logger.info("Using Adam to craft poison samples")
+            optimizer = torch.optim.Adam(poison_batch.parameters(), lr=self.learning_rate, betas=(self.momentum, 0.999))
+
+        base_tensor_batch = torch.stack(base_tensor_list, 0)
+        base_range01_batch = base_tensor_batch * std + mean
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:168]: Does this step un-normalise `x`?

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+from art.estimators.classification.classifier import ClassifierMixin
+from art.estimators.classification.pytorch import PyTorchClassifier
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class BullseyePolytopeAttack(PoisoningAttackWhiteBox):
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:44]:
```suggestion
class BullseyePolytopeAttackPyTorch(PoisoningAttackWhiteBox):
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+"""
+This module implements clean-label attacks on Neural Networks.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+from typing import Optional, Tuple, Union, TYPE_CHECKING, List
+
+import numpy as np
+import torch
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:27]:
```suggestion
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+import math
+import time
+from torch.optim.optimizer import Optimizer
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:30]:
```suggestion
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements clean-label attacks on Neural Networks.
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:19]: Is this module docstring still correct?

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+                           with varying architectures.
+        :param target: The target input(s) of shape (N, W, H, C) to misclassify at test time. Multiple targets will be
+                       averaged.
+        :param feature_layer: The name(s) of the feature representation layer(s).
+        :param opt: The optimizer to use for the attack. Can be 'adam', 'sgd', or 'signedadam'
+        :param max_iter: The maximum number of iterations for the attack.
+        :param learning_rate: The learning rate of clean-label attack optimization.
+        :param momentum: The momentum of clean-label attack optimization.
+        :param decay_iter: Which iterations to decay the learning rate.
+                           Can be a integer (every N interations) or list of integers [0, 500, 1500]
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:103]:
```suggestion
                           Can be a integer (every N iterations) or list of integers [0, 500, 1500]
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+from torch.optim.optimizer import Optimizer
+from tqdm.auto import trange
+
+from art.attacks.attack import PoisoningAttackWhiteBox
+from art.estimators import BaseEstimator, NeuralNetworkMixin
+from art.estimators.classification.classifier import ClassifierMixin
+from art.estimators.classification.pytorch import PyTorchClassifier
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:39]:
```suggestion
if TYPE_CHECKING:
    import torch
    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
```

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+        """
+        Iteratively finds optimal attack points starting at values at x
+
+        :param x: The base images to begin the poison process.
+        :param y: Target label
+        :param mean: The mean of the dataset. Defaults to CIFAR10 Mean.
+        :param std: The standard deviation of the dataset. Defaults to CIFAR std
+        :return: An tuple holding the (poisoning examples, poisoning labels).
+        """
+        if mean is None:
```
Comment (ng390) [inline art/attacks/poisoning/bullseye_polytope_attack.py:147]: Is it possible to allow mean to be a list, which can be cast to a torch Tensor? Could include a check to ensure the list was the correct length. Would make armory integration cleaner, especially with configs.

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+                                          self.endtoend, self.feature_layer)
+            total_loss.backward()
+
+            optimizer.step()
+
+            # clip the perturbations into the range
+            perturb_range01 = torch.clamp((poison_batch.poison.data - base_tensor_batch) * std,
+                                          -self.epsilon,
+                                          self.epsilon)
+            perturbed_range01 = torch.clamp(base_range01_batch.data + perturb_range01.data, 0, 1)
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:213]: Can we use here `estimator.clip_values` for clipping?

File: tests/attacks/poison/test_bullseye_polytope_attack.py (hunk 4 above)
Comment (ebubae) [reply tests/attacks/poison/test_bullseye_polytope_attack.py:39]: Added in commit

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 5 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Done

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 14 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Done.

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 5 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Done.

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 7 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Done

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 10 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Fixed.

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 11 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Thanks for catching this. Corrected it now.

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 14 above)
Comment (ebubae) [reply art/attacks/poisoning/bullseye_polytope_attack.py]: Done.

File: art/attacks/poisoning/bullseye_polytope_attack.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+        self.max_iter = max_iter
+        self.verbose = verbose
+        self._check_params()
+
+    def poison(
+            self,
+            x: np.ndarray,
+            y: Optional[np.ndarray] = None,
+            mean: Optional[List[float]] = None,
+            std: Optional[List[float]] = None,
```
Comment (beat-buesser) [inline art/attacks/poisoning/bullseye_polytope_attack.py:132]: Can we remove these arguments and replace them with the estimators preprocessing instead?

File: art/attacks/poisoning/bullseye_polytope_attack.py (hunk 13 above)
Comment (beat-buesser) [reply art/attacks/poisoning/bullseye_polytope_attack.py:159]: We have to remove the `mean` and `std` from the method `poison` and add them to the classifier's preprocessing instead.
